# Variant analysis workshop


## Dataset Info and TODO:
- 1 Trio for preprocessing
    - two extractions... one that is big (all of the data) and one that is small
    - small dataset for the line by line code they (50KB region from the bam file that has SNPS/indels) 
    - check out how many reads/size of two chromosomes as the big extraction (then subset reads if necessary)
- All data from VCF files I already have for Plink stuff (just subset chroms)
- Find a paper similar to make sure we are covering everything properly
- See if there is any more metadata to stay away from cranio questions and genes

## Mon:


CLI/R

Cluster

## Tue: 

Finish CLI/R 
- something

Lunch Break

Data Reduction
- Fastq file discusion
- Running HTS (Brad)
- HTStream QA/QC MultiQC (Brad)
- samtools/bwa
- Bam file discussion
- IGG examples, variants in IGG
- Visualize trio in IGG to identify variant and inheritance 

## Wed:

Data Reduction
- GATK,FreeBayes,DeepVariant
- 

Lunch Break

Data Reduction
- VCF file discussion
- Comparisons of FreeBayes/GATK/DeepVariant


## Thur:

Data Analysis
- Plink 
- Normal association vs TDT
- 

Anthony Talk?
- Try to rope him in to more

Exome Talk?

Lunch Break

Data Analysis
- Plink 
- IBD
- Annotation with wAnnovar

## Fri:

R visualization of data

Manhattan plots 

PlinkQC library

Bioinformatics talk

Experimental Design